gitlab/client.py

```diff
@@ -97,13 +100,15 @@ def __init__(
         self.http_password = http_password
         self.oauth_token = oauth_token
         self.job_token = job_token
-        self._set_auth_info()
+        self.oauth_credentials = oauth_credentials
 
         #: Create a session object for requests
         http_backend: Type[http_backends.DefaultBackend] = kwargs.pop(
             "http_backend", http_backends.DefaultBackend
         )
         self.http_backend = http_backend(**kwargs)
+
+        self._set_auth_info()
```

Comment:
We have to configure the backend first to be able to make requests in `_set_auth_info()` because that potentially makes a request to retrieve the OAuth token, so moving this down here.

Reply:
It seems like client is a monolithic class.  Moving auth properties and functions to an auth class would be easier to maintain. IMO, implementing the auth class should come first before switching to a different method.